File: Backend/rest/src/main/java/net/chrisparton/sparkled/rest/json/UnsignedByteArrayTypeAdapter.java
package net.chrisparton.sparkled.rest.json;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;

/**
 * Serialises bytes as unsigned (0-255) instead of signed. Code producing the bytes that will be serialised must take
 * take this conversion into consideration.
 */
public class UnsignedByteArrayTypeAdapter implements JsonSerializer<byte[]> {

    public JsonElement serialize(byte[] src, Type typeOfSrc, JsonSerializationContext context) {
        JsonArray array = new JsonArray();
        for (byte b : src) {
            array.add(b & 0xFF);
        }
        return array;
    }
}

File: model/src/main/kotlin/io/sparkled/model/util/UnsignedByteArrayTypeAdapter.kt
package io.sparkled.model.util

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer

import java.lang.reflect.Type

/**
 * Serialises bytes as unsigned (0-255) instead of signed (-128-127). Code producing the bytes that will be serialised
 * must take this conversion into consideration.
 */
class UnsignedByteArrayTypeAdapter : JsonSerializer<ByteArray> {

    fun serialize(src: ByteArray, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val array = JsonArray()
        for (b in src) {
            array.add(b and 0xFF)
        }
        return array
    }
}
